lib/color_spinor_field.cpp

```diff
@@ -64,41 +79,150 @@ namespace quda
+    nColor(std::exchange(field.nColor, 0)),
+    nSpin(std::exchange(field.nSpin, 0)),
+    nVec(std::exchange(field.nVec, 0)),
+    twistFlavor(std::exchange(field.twistFlavor, QUDA_TWIST_INVALID)),
+    pc_type(std::exchange(field.pc_type, QUDA_PC_INVALID)),
+    suggested_parity(std::exchange(field.suggested_parity, QUDA_INVALID_PARITY)),
+    length(std::exchange(field.length, 0)),
+    v(std::exchange(field.v, nullptr)),
+    v_h(std::exchange(field.v_h, nullptr)),
+    norm_offset(std::exchange(field.norm_offset, 0)),
+    ghost(std::exchange(field.ghost, { })),
+    ghostFace(std::exchange(field.ghostFace, { })),
+    ghostFaceCB(std::exchange(field.ghostFaceCB, { })),
+    ghost_buf(std::exchange(field.ghost_buf, { })),
+    dslash_constant(std::exchange(field.dslash_constant, nullptr)),
+    bytes(std::exchange(field.bytes, 0)),
+    bytes_raw(std::exchange(field.bytes_raw, 0)),
+    siteOrder(std::exchange(field.siteOrder, QUDA_INVALID_SITE_ORDER)),
+    fieldOrder(std::exchange(field.fieldOrder, QUDA_INVALID_FIELD_ORDER)),
+    gammaBasis(std::exchange(field.gammaBasis, QUDA_INVALID_GAMMA_BASIS)),
+    even(std::exchange(field.even, nullptr)),
+    odd(std::exchange(field.odd, nullptr)),
+    composite_descr(std::exchange(field.composite_descr, CompositeColorSpinorFieldDescriptor())),
+    components(std::move(field.components))
+  { }
+
   ColorSpinorField::~ColorSpinorField()
   {
-    if (dslash_constant) host_free(dslash_constant);
     destroy();
-    if (Location() == QUDA_CUDA_FIELD_LOCATION) destroyComms();
+  }
+
+  void ColorSpinorField::clear()
+  {
+    LatticeField::clear();
+    destroy();
+  }
+
+  ColorSpinorField &ColorSpinorField::operator=(const ColorSpinorField &src)
+  {
+    if (&src != this) {
+      if (!init) { // keep current attributes unless unset
+        LatticeField::operator=(src);
+        composite_descr = src.composite_descr;
+        ColorSpinorParam param;
+        src.fill(param);
+        param.create = QUDA_COPY_FIELD_CREATE;
+        create(param);
+      }
+
+      copy(src);
+    }
+    return *this;
+  }
+
+  static bool are_compatible(ColorSpinorField &a, ColorSpinorField &b)
+  {
+    bool rtn = true;
+    if (a.Precision() != b.Precision()   ||
+        a.FieldOrder() != b.FieldOrder() ||
+        a.SiteSubset() != b.SiteSubset() ||
+        a.VolumeCB() != b.VolumeCB()     ||
+        a.Ncolor() != b.Ncolor()         ||
+        a.Nspin() != b.Nspin()           ||
+        a.Nvec() != b.Nvec()             ||
+        a.TwistFlavor() != b.TwistFlavor())
+      rtn = false;
+
+    return rtn;
+  }
+
+  ColorSpinorField &ColorSpinorField::operator=(ColorSpinorField &&src)
+  {
+    if (&src != this) {
+      // if field not already initialized then move the field
+      if (!init || are_compatible(*this, src)) {
+        if (init) destroy();
+        LatticeField::operator=(std::move(src));
+
+        init = std::exchange(src.init, false);
```

Comment:
replicating the code for move seems error prone if we ever touch that code.

Reply:
I could make a `move` member function that is called by both the move constructor and assignment operator.  Similar in spirit to the `create` function.  Shall I do that?

Reply:
Yeah, I guess. I believe this is better than duplicating the code explicitly. Unless someone knows some magic function that makes a default move constructor work here magically. Like "please dear compiler, could you please cal std:Exchange for all my members?"

Reply:
340a1d37de089d3abeb813d5edbb08499e02945d